### ISSUE TITLE
Fix /getdata failing on V1

### ISSUE THREAD

**Issue by wookhuk** [zenvisage/zenvisage#23]

**Comment by eidehua** [comment]
So we seem to be using the /getdata path for the existing trends as well as the main submit path. Not sure what to change the existing trends to, but for the main one (the function called getData(query, xAxisType, yAxisType) ) we can use the xAxis and yAxis type to call to the correct backend call. But we might need to refactor the callback function

**Comment by superchaoran** [comment]
Fixed.

**Comment by superchaoran** [comment]
The built-in Apache Euclidean distance is using larger array's index for small array accessing which caused the bug.

**Comment by eidehua** [comment]
A new bug is breaking getdata again. I've checked the json we are passing to the backend, seems like there is a new "xRange":[1,46] key value pair in the chi version? What would this correspond to in the V1 version?

Caused by:
java.lang.NullPointerException
        at edu.uiuc.zenvisage.service.Similarity.computeOrders(Similarity.java:137)
        at edu.uiuc.zenvisage.service.Similarity.compute(Similarity.java:102)
        at edu.uiuc.zenvisage.service.ZvMain.runDragnDropInterfaceQuery(ZvMain.java:259)
        at edu.uiuc.zenvisage.api.ZvBasicAPI.getData(ZvBasicAPI.java:54)

**Comment by eidehua** [comment]
The "x" in v1 corresponds to pixels right now
`{"method":"SimilaritySearch","yAxis":"SoldPrice","xAxis":"Quarter","groupBy":"Metro","aggrFunc":"avg","aggrVar":"SoldPrice","outlierCount":50,"dataX":[],"dataY":[],"yMax":140.82678125,"yMin":103.5094453125,"sketchPoints":[{"points":[{"x":0,"y":244.5},{"x":32,"y":244.5},{"x":64,"y":244.5},{"x":96,"y":244.5},{"x":128,"y":244.5},{"x":160,"y":244.5},{"x":192,"y":244.5},{"x":224,"y":244.5},{"x":256,"y":244.5},{"x":288,"y":244.5},{"x":320,"y":244.5},{"x":352,"y":244.5},{"x":384,"y":244.5},{"x":416,"y":244.5},{"x":448,"y":244.5},{"x":480,"y":244.5},{"x":512,"y":244.5},{"x":544,"y":244.5},{"x":576,"y":244.5},{"x":608,"y":244.5},{"x":640,"y":244.5},{"x":672,"y":244.5},{"x":704,"y":244.5},{"x":736,"y":244.5},{"x":768,"y":244.5},{"x":800,"y":244.5},{"x":832,"y":244.5},{"x":864,"y":244.5},{"x":896,"y":244.5},{"x":928,"y":244.5}],"minX":0,"maxX":977,"minY":0,"maxY":489,"yAxis":"SoldPrice","xAxis":"Quarter","groupBy":"Metro","aggrFunc":"avg","aggrVar":"SoldPrice"}],"distanceNormalized":true,"outputNormalized":true,"clustering":"KMeans","distance_metric":"Euclidean","predicateOperator":"=","predicateColumn":"Metro","predicateValue":""}`


While the "x" in chi is just index?
`{"method":"SimilaritySearch","xAxis":"Quarter","yAxis":"SoldPrice","groupBy":"Metro","aggrFunc":"avg","aggrVar":"SoldPrice","outlierCount":"50","dataX":[],"dataY":[],"yMax":null,"yMin":null,"sketchPoints":[{"points":[{"x":1,"y":119736.9609375},{"x":2,"y":126123.8515625},{"x":3,"y":129926.5625},{"x":4,"y":129576.875},{"x":5,"y":128102.9140625},{"x":6,"y":130122.2890625},{"x":7,"y":136658.328125},{"x":8,"y":137745.828125},{"x":9,"y":139802.75},{"x":10,"y":139066.25},{"x":11,"y":145962.921875},{"x":12,"y":143427.078125},{"x":13,"y":140925},{"x":14,"y":142875},{"x":15,"y":146854.171875},{"x":16,"y":145900.78125},{"x":17,"y":143327.34375},{"x":18,"y":147202.078125},{"x":19,"y":150774.28125},{"x":20,"y":146928.578125},{"x":21,"y":138607.546875},{"x":22,"y":138114.140625},{"x":23,"y":143456.25},{"x":24,"y":141320.65625},{"x":25,"y":142064.359375},{"x":26,"y":137791.796875},{"x":27,"y":146974.4375},{"x":28,"y":142168.90625},{"x":29,"y":132104.984375},{"x":30,"y":142323.484375},{"x":31,"y":148512.015625},{"x":32,"y":144656.25},{"x":33,"y":144695.515625},{"x":34,"y":144794.734375},{"x":35,"y":144717.703125},{"x":36,"y":137936.453125},{"x":37,"y":144800.6875},{"x":38,"y":145630.828125},{"x":39,"y":153156.984375},{"x":40,"y":146146.125},{"x":41,"y":141132.234375},{"x":42,"y":148655.96875},{"x":43,"y":154421.703125},{"x":44,"y":147481.515625},{"x":45,"y":144498.796875},{"x":46,"y":152623.75}],"minX":1,"maxX":46,"minY":15384.615,"maxY":33179964,"yAxis":"SoldPrice","xAxis":"Quarter","groupBy":"Metro","aggrFunc":"avg","aggrVar":"SoldPrice"}],"distanceNormalized":false,"outputNormalized":false,"clustering":"KMeans","predicateOperator":"=","predicateColumn":"Metro","predicateValue":"","xRange":[1,46]}`

**Comment by adityagp** [comment]
@eidehua -- thanks for the issue.  can you coordinate with @wookhuk and @siddta on fixing and whether it makes sense to converge the two implementations?

**Comment by wookhuk** [comment]
@eidehua There are a lot of changes going on in the backend which will break v1. The drawing system is completely different and there is no way one backend can handle this change. We forked the code to keep things separate, but we ended up merging so v1 can benefit from some improvements. But I don't think this is sustainable with two completely different frontends. 

TLDR: Two options. 
1) Fork to keep things separate and revert this branch to a version where it doesn't break. CHI version development will continue on a different fork.
2) Abandon V1.

**Comment by adityagp** [comment]
@wookhuk and @siddta discussed this and we think it would be good to continue without forking, while trying to keep things as general as possible (in order to not break v1 any more than absolutely necessary).

**Comment by siddta** [comment]
Either of the version uses almost the same set of api calls to feed the visualizations. Over the past few days, we have only added a minor improvements to the backend to do partial similarity search. 

Sending pixels or domain value to the backend is the only difference, I guess? I will look at both the implementations and see how we can resolve this hopefully later today or tomorrow.

@wookhuk: what other changes are needed at the backend? could we create issues for those or discuss sometime?

**Comment by eidehua** [comment]
Yeah, I feel keeping the backend as general as possible should be best, no matter the v1 and chi issue. Like someone else could develop their own front-end using our backend calls. 

The drawing systems are completely different but as long as we process it into a format the backend understands, it should be good. I'm sure V1 can also use the multiple backend calls instead of just /getdata as well.

V1 I feel can be left as is feature wise, just bug fixing, and then when chi has even more features than V1 we can have that as the main frontend?

**Comment by siddta** [comment]
Yes, agreed. There is not much difference between the two versions at this stage. Major functionalities and Interactions are all same between the two version. CHI one uses dygraph for drawing, other than that I would say we make effort to understand the code and keep everything consistent by reusing/refactoring/adding. I am ok if V1 remains in a broken state for a few days (chances are less once we fix the domain pixel thing).


I see currently we are checking into so many versions. We need to merge all versions into the master version. If there is conflict, chi functionality should override V1's. But please create an issue so that all know what might not be working inaster.
